src/progress_bar_exercise/components/ProgressBar/useProgressBarState.js

```diff
@@ -0,0 +1,118 @@
+import { useCallback, useEffect, useMemo, useReducer, useRef } from "react";
+
+export const DEFAULT_SECONDS = 15;
+export const ONE_SECOND = 1000;
+const PERCENTAGE_STOP = 90;
+
+const ACTIONS = {
+  reset: "RESET",
+  addNormalAnimation: "ADD_NORMAL_ANIMATION",
+  setTime: "SET_TIME",
+  done: "DONE",
+};
+
+const initialState = {
+  passedMilliseconds: 0,
+  transitionDuration: 0,
+  realProgress: 0,
+  animatedProgress: 0,
+  hasStarted: false,
+};
+
+function init(initialState) {
+  const expectedMilliseconds = initialState.expectedSeconds * ONE_SECOND;
+  return {
+    ...initialState,
+    expectedMilliseconds,
+    stopMilliseconds: expectedMilliseconds * (PERCENTAGE_STOP / 100.0),
+  };
+}
+
+const getProgress = (passedTime, totalTime) => {
+  return Math.round((passedTime / totalTime) * 100);
+};
+
+function reducer(state, action) {
```

Comment:
**Note for reviewers:**

Even if `useReducer` can feel more complex at times, I noticed that in this case we need to update many pieces of state at once and that's easier with a reducer than with `useState`